selfservice/strategy/code/strategy.go

```diff
@@ -82,14 +118,319 @@ func NewStrategy(deps strategyDependencies) *Strategy {
+// this is used with the `recovery`, `verification`, `registration` and `login` flows.
+func (s *Strategy) NewCodeUINodes(r *http.Request, f flow.Flow, data json.RawMessage) error {
+	if err := s.PopulateMethod(r, f); err != nil {
+		return err
+	}
+
+	// on Registration flow we need to populate the form with the values from the initial form generation
+	if f.GetFlowName() == flow.RegistrationFlow {
+		for _, n := range container.NewFromJSON("", node.CodeGroup, data, "traits").Nodes {
+			// we only set the value and not the whole field because we want to keep types from the initial form generation
+			f.GetUI().GetNodes().SetValueAttribute(n.ID(), n.Attributes.GetValue())
+		}
+	} else if f.GetFlowName() == flow.LoginFlow {
+		// on Login flow we need to populate the form with the values from the initial form generation
+		for _, n := range container.NewFromJSON("", node.DefaultGroup, data, "").Nodes {
+			f.GetUI().GetNodes().SetValueAttribute(n.ID(), n.Attributes.GetValue())
+		}
+	}
+
+	return nil
+}
+
+type (
+	CreateCodeState           func(context.Context, *CodeStateManagerPayload) error
+	ValidateCodeState         func(context.Context, *CodeStateManagerPayload) error
+	AlreadyValidatedCodeState func(context.Context, *CodeStateManagerPayload) error
+	CodeStateManager          struct {
+		f                         flow.Flow
+		payload                   *CodeStateManagerPayload
+		createCodeState           CreateCodeState
+		verifyCodeState           ValidateCodeState
+		alreadyValidatedCodeState AlreadyValidatedCodeState
+	}
+	CodeStateManagerPayload struct {
+		Identifier       string
+		Email            string
+		Traits           json.RawMessage
+		TransientPayload json.RawMessage
+		Resend           string
+		Code             string
+	}
+)
+
+func NewCodeStateManager(f flow.Flow, payload *CodeStateManagerPayload) *CodeStateManager {
+	return &CodeStateManager{
+		f:       f,
+		payload: payload,
+	}
+}
+
+func (c *CodeStateManager) RegisterCreateCodeState(fn CreateCodeState) {
+	c.createCodeState = fn
+}
+
+func (c *CodeStateManager) RegisterVerifyCodeState(fn ValidateCodeState) {
+	c.verifyCodeState = fn
+}
+
+func (c *CodeStateManager) RegisterAlreadyValidatedCodeState(fn AlreadyValidatedCodeState) {
+	c.alreadyValidatedCodeState = fn
+}
+
+func (c *CodeStateManager) validatePayload(ctx context.Context) error {
+	switch c.f.GetFlowName() {
+	case flow.LoginFlow:
+		if len(c.payload.Identifier) == 0 {
+			return errors.WithStack(schema.NewRequiredError("#/identifier", "identifier"))
+		}
+	case flow.RegistrationFlow:
+		if len(c.payload.Traits) == 0 {
+			return errors.WithStack(schema.NewRequiredError("#/traits", "traits"))
+		}
+	case flow.RecoveryFlow, flow.VerificationFlow:
+		if len(c.payload.Email) == 0 {
+			return errors.WithStack(schema.NewRequiredError("#/email", "email"))
+		}
+	default:
+		return errors.New("received unexpected flow type")
+	}
+	return nil
+}
+
+func (c *CodeStateManager) Run(ctx context.Context) error {
```

Comment:
function to help run the code method logic.

works with login and registration. can also be adopted in recovery and verification.